File: models_py/m5_k_poisson.py
"""M5 - strikeouts per start (Python side of the parity gate).

Poisson regression of strikeout count on batters faced -> expected K totals per start
(prop-bet framing). Reads ONLY the gold feed; writes a flat result file.
"""
from __future__ import annotations

import json
import os

import pandas as pd
import statsmodels.api as sm


def main() -> None:
    gold = os.getenv("MLB_GOLD_DIR", "data/gold")
    res = os.getenv("MLB_RESULTS_DIR", "data/results")
    os.makedirs(res, exist_ok=True)

    df = pd.read_parquet(f"{gold}/pitcher_start.parquet")
    x = sm.add_constant(df[["bf"]].to_numpy())
    model = sm.GLM(df["k"].to_numpy(), x, family=sm.families.Poisson()).fit()  # log link, IRLS

    out = {
        "n": int(len(df)),
        "coef": {
            "intercept": round(float(model.params[0]), 8),
            "bf": round(float(model.params[1]), 8),
        },
        "mean_k_per_start": round(float(df["k"].mean()), 4),
        "k_per_bf": round(float(df["k"].sum() / df["bf"].sum()), 6),
    }
    with open(f"{res}/m5_k_poisson__py.json", "w") as f:
        json.dump(out, f, indent=2)
    print(json.dumps({"stage": "models_py", "model": "m5_k_poisson", **out}))


if __name__ == "__main__":
    main()

File: models_r/m5_k_poisson.R
# M5 - strikeouts per start (R side of the parity gate).
# Poisson regression of strikeout count on batters faced -> expected K totals per start.
suppressMessages({
  library(arrow)
  library(dplyr)
  library(jsonlite)
})

gold <- Sys.getenv("MLB_GOLD_DIR", "data/gold")
res <- Sys.getenv("MLB_RESULTS_DIR", "data/results")
dir.create(res, showWarnings = FALSE, recursive = TRUE)

df <- read_parquet(file.path(gold, "pitcher_start.parquet"))
m <- glm(k ~ bf, data = df, family = poisson())

co <- coef(m)
out <- list(
  n = nrow(df),
  coef = list(
    intercept = round(co[["(Intercept)"]], 8),
    bf = round(co[["bf"]], 8)
  ),
  mean_k_per_start = round(mean(df$k), 4),
  k_per_bf = round(sum(df$k) / sum(df$bf), 6)
)

# digits = 12: preserve precision for EXACT parity (see DECISIONS.md).
write_json(out, file.path(res, "m5_k_poisson__r.json"), auto_unbox = TRUE, pretty = TRUE, digits = 12)
cat(toJSON(c(list(stage = "models_r", model = "m5_k_poisson"), out), auto_unbox = TRUE, digits = 12), "\n")
